addons/account/wizard/account_invoice_send.py

```diff
@@ -134,6 +125,16 @@ def _send_email(self):
                 if prioritary_attachments:
                     main_attachment = prioritary_attachments[0]
                     invoice.with_context(tracking_disable=True).sudo().write({'message_main_attachment_id': main_attachment.id})
```

Comment:
Isn't the write delayed by the ORM anyway ? So, to me, just moving the code sending the mails below it will not really solve the issue, unless you flush ; will it ? :thinking:

Reply:
The .write call is what might cause the concurrency issue due to the compute chain, and we can catch it in `retrying()` within [model.py](https://github.com/odoo/odoo/blob/a433f3175a5105d9028dd9937d5461ae2d5d11fe/odoo/service/model.py#L137): 

Calling a flush is redundant, since we will catch this error within `retrying()`.

The current code might retry the origin func (in this case, `call_button()`, which calls `send_and_print_action()`), by 1) creating and sending the emails first (which does not have a concurrency issue) -> 2) do the attachment write which may find a concurrency issue. The result of this is that multiple rounds of A) may be run before B) finishes without errors. 

Moving the mail creation and sending below the attachment write call resolves this issue because the _send_email is synchronous and doesn't branch; we will not call what follows the attachment write until there is no concurrency issue found in resulting compute chain, which then avoids the mail creation and sending. In other words, 2) must succeed before 1) is called. 

More details regarding this investigation are found in the chatter of the task (but might contain sensitive information), let me know if you have other questions or concerns.